pkg/migration-tool/utils/client.go

```diff
@@ -85,22 +85,32 @@ func NewClient(remoteName, urlString string, clientType uint, timeout model.Dura
 	}, nil
 }
 
+// PrompbResponse is a type that contains promb-result and information pertaining to it.
+type PrompbResponse struct {
+	ID                   int
+	Result               *prompb.QueryResult
+	NumBytesCompressed   int
+	NumBytesUncompressed int
+}
+
 // Read reads from a remote endpoint. It returns the response size of compressed and uncompressed in bytes.
-func (c *Client) Read(ctx context.Context, query *prompb.Query, desc string) (result *prompb.QueryResult, numBytesCompressed int, numBytesUncompressed int, err error) {
+func (c *Client) Read(ctx context.Context, query *prompb.Query, shardID int, desc string, doneChan chan<- *PrompbResponse, errChan chan<- error) {
```

Comment:
As I mentioned before, I'd keep Read as it was and just add a ReadChannels that responds on channels.

Reply:
Agreed. That would be much better.